### DISCUSSION TITLE
Graph support for rich

### DISCUSSION THREAD

**Discussion by LarsHill** [Textualize/rich#1074] [Ideas]
In addition to the great tree rendering option, it would be cool to have some simple graph visualization capabilities. 

Maybe one could start supporting directed graphs in a hierarchical sugiyama layout. 
Basically this would be an extension of tree rendering (top to bottom instead of left to right) with the option to connect multiple predecessor nodes to a singe ancestor node.

Would something like this be feasible to include into rich?

I think it would nicely complement the feature palette of tree and table rendering. 

Best regards
Lars

**Comment by willmcgugan** [comment]
Sure it would be nice, but that would be a challenge to render to the terminal and I only have so much time. It doesn't feel as universally applicable as the current feature set so I can't see it going in the core. It is my hope that third parties will implement such things using the [render protocol](https://rich.readthedocs.io/en/latest/protocol.html).

**Reply by LarsHill** [reply]
That makes sense. I just noticed that [DVC](https://dvc.org/doc/command-reference/dag) has console graph visualization functionality build in.

They use [grandalf](https://github.com/bdcht/grandalf) to create the graph sugiyama layout and wrote their own ascii rendering [script](https://github.com/iterative/dvc/blob/master/dvc/dagascii.py) to visualize the graph in the console.

I‘ll take a look at adapting their script to use rich for the rendering part.
